network/p2p/tracer/gossipSubMeshTracer.go

```diff
@@ -0,0 +1,184 @@
+package tracer
+
+import (
+	"sync"
+	"time"
+
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/network/p2p"
+	"github.com/onflow/flow-go/utils/logging"
+)
+
+const (
+	// MeshLogIntervalMsg is the message logged by the tracer every logInterval.
+	MeshLogIntervalMsg = "topic mesh peers of local node since last heartbeat"
+)
+
+// GossipSubMeshTracer is a tracer that tracks the local mesh peers for each topic.
+// It also logs the mesh peers and updates the local mesh size metric.
+type GossipSubMeshTracer struct {
+	component.Component
+	pubsub.RawTracer
+
+	topicMeshMu    sync.RWMutex                    // to protect topicMeshMap
+	topicMeshMap   map[string]map[peer.ID]struct{} // map of local mesh peers by topic.
+	logger         zerolog.Logger
+	idProvider     module.IdentityProvider
+	loggerInterval time.Duration
+	metrics        module.GossipSubLocalMeshMetrics
+}
+
+var _ p2p.PubSubTracer = (*GossipSubMeshTracer)(nil)
+
+func NewGossipSubMeshTracer(
+	logger zerolog.Logger,
+	metrics module.GossipSubLocalMeshMetrics,
+	idProvider module.IdentityProvider,
+	loggerInterval time.Duration) *GossipSubMeshTracer {
+
+	g := &GossipSubMeshTracer{
+		RawTracer:      NewGossipSubNoopTracer(),
+		topicMeshMap:   make(map[string]map[peer.ID]struct{}),
+		idProvider:     idProvider,
+		metrics:        metrics,
+		logger:         logger.With().Str("component", "gossip_sub_topology_tracer").Logger(),
+		loggerInterval: loggerInterval,
+	}
+
+	g.Component = component.NewComponentManagerBuilder().
+		AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+			ready()
+			g.logLoop(ctx)
+		}).
+		Build()
+
+	return g
+}
+
+// GetMeshPeers returns the local mesh peers for the given topic.
+func (t *GossipSubMeshTracer) GetMeshPeers(topic string) []peer.ID {
+	t.topicMeshMu.RLock()
+	defer t.topicMeshMu.RUnlock()
+
+	peers := make([]peer.ID, 0, len(t.topicMeshMap[topic]))
+	for p := range t.topicMeshMap[topic] {
+		peers = append(peers, p)
+	}
+	return peers
+}
+
+// Graft is called when a peer is added to a topic mesh. The tracer uses this to track the mesh peers.
+func (t *GossipSubMeshTracer) Graft(p peer.ID, topic string) {
+	t.topicMeshMu.Lock()
+	defer t.topicMeshMu.Unlock()
+
+	lg := t.logger.With().Str("topic", topic).Str("peer_id", p.String()).Logger()
+
+	if _, ok := t.topicMeshMap[topic]; !ok {
+		t.topicMeshMap[topic] = make(map[peer.ID]struct{})
+	}
+	t.topicMeshMap[topic][p] = struct{}{}
+
+	id, exists := t.idProvider.ByPeerID(p)
+	if !exists {
+		lg.Warn().
+			Bool(logging.KeySuspicious, true).
+			Msg("grafted peer not found in identity provider")
+		return
```

Comment:
do we still need to update the metric since its size may have changed?

Reply:
Size changes only when a `GRAFT` or `PRUNE` message is sent or received. In both cases, their [relative call-back tracer methods]https://github.com/onflow/flow-go/blob/b6e255a84748b55a4de6c68053194ff9fb1332bc/network/p2p/tracer/gossipSubMeshTracer.go#L80-L132() are called by the GossipSub and the size metric is updated. The periodic logging is primarily for the situations where the topic-meshes remain the same with no new GRAFT and PRUNE sent or received. Hence, we don't have any event-based logging through `GRAFT` and `PRUNE` method calls.